sway-lsp/src/core/session.rs

```diff
@@ -165,12 +165,71 @@ impl Session {
         sync: &SyncWorkspace,
     ) -> Option<GotoDefinitionResponse> {
         let _p = tracing::trace_span!("token_definition_response").entered();
-        token_map
-            .token_at_position(uri, position)
-            .and_then(|item| item.value().declared_token_ident(engines))
+        let token_at_pos = token_map.token_at_position(uri, position)?;
+        let token = token_at_pos.value();
+        // If this is a type parameter, check if we're inside a struct declaration and restrict lookup
+        if let Some(crate::core::token::TypedAstToken::TypedParameter(_)) = token.as_typed() {
+            // Find the parent struct declaration token that contains this range
+            let param_range = &token_at_pos.key().range;
+            // Limit the number of struct tokens checked to prevent stack overflow
+            let mut struct_token = None;
+            let mut checked = 0;
+            for entry in token_map.tokens_for_file(uri) {
+                if checked > 1000 {
```

Comment:
This arbitrary limit of 1000 iterations has no justification and could break on legitimate large files - either fix the underlying O(n²) complexity or provide proper rationale for this number.